lib/rpmug.c

```diff
@@ -1,16 +1,124 @@
 #include "system.h"
 
-#include <pthread.h>
-#include <pwd.h>
-#include <grp.h>
-#include <netdb.h>
+#include <errno.h>
 #include <rpm/rpmlog.h>
 #include <rpm/rpmstring.h>
+#include <rpm/rpmmacro.h>
 
 #include "lib/misc.h"
 #include "lib/rpmug.h"
 #include "debug.h"
 
+static char *pwpath = NULL;
+static char *grppath = NULL;
+
+static const char *getpath(const char *bn, const char *dfl, char **dest)
+{
+    if (*dest == NULL) {
+	char *s = rpmExpand("%{_", bn, "_path}", NULL);
+	if (*s == '%' || *s == '\0') {
+	    free(s);
+	    s = xstrdup(dfl);
+	}
+	*dest = s;
+    }
+    return *dest;
+}
+
+static const char *pwfile(void)
+{
+    return getpath("passwd", "/etc/passwd", &pwpath);
+}
+
+static const char *grpfile(void)
+{
+    return getpath("group", "/etc/group", &grppath);
+}
+
+/*
+ * Lookup an arbitrary field based on contents of another in a ':' delimited
+ * file, such as /etc/passwd or /etc/group.
+ */
+static int lookup_field(const char *path, const char *val, int vcol, int rcol,
+			char **ret)
+{
+    int rc = -1; /* assume not found */
+    char *str, buf[BUFSIZ];
+    FILE *f = fopen(path, "r");
+    if (f == NULL) {
+	rpmlog(RPMLOG_ERR, _("failed to open %s for id/name lookup: %s\n"),
+		path, strerror(errno));
+	return rc;
+    }
+
+    while ((str = fgets(buf, sizeof(buf), f)) != NULL) {
+	int nf = vcol > rcol ? vcol : rcol;
+	const char *fields[nf + 1];
+	char *tok, *save = NULL;
+	int col = -1;
+
+	while ((tok = strtok_r(str, ":", &save)) != NULL) {
+	    fields[++col] = tok;
+	    str = NULL;
+	    if (col >= nf)
+		break;
+	}
+
+	if (col >= nf) {
+	    if (rstreq(val, fields[vcol])) {
+		*ret = xstrdup(fields[rcol]);
+		rc = 0;
+	    }
+	}
+    }
+
+    fclose(f);
+
+    return rc;
+}
+
+/* atol() with error handling, return 0/-1 on success/failure */
+static int stol(const char *s, long *ret)
+{
+    int rc = 0;
+    char *end = NULL;
+    long val = strtol(s, &end, 10);
+
+    /* only accept fully numeric data */
+    if (*s == '\0' || *end != '\0')
+	rc = -1;
+
+    if ((val == LONG_MIN || val == LONG_MAX) && errno == ERANGE)
+	rc = -1;
+
+    if (rc == 0)
+	*ret = val;
+
+    return rc;
+}
+
+static int lookup_num(const char *path, const char *val, int vcol, int rcol,
+			long *ret)
+{
+    char *buf = NULL;
+    int rc = lookup_field(path, val, vcol, rcol, &buf);
+    if (rc == 0) {
+	rc = stol(buf, ret);
+	free(buf);
+    }
+    return rc;
+}
+
+static int lookup_name(const char *path, long val, int vcol, int rcol,
```

Comment:
A nitpick but what about `lookup_str()` as a function name instead, for the sake of consistency? :smile: (The function doesn't do anything specific to "names")

Reply:
Might be a nit, but it reasonably fair one :smile: